### PR TITLE
Spark: Refactor building write distribution and ordering

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR has the following contributions:

- When I was working on distributions for row-level commands, I realized that it is easier to decide which distribution mode to use  inside `SparkWriteConf`. Right now, we only parse the config in `SparkWriteConf` but then have extra checks in the util while building the actual distribution. That means we have bits of the same logic in multiple places.
- We no longer need `OrderField`. We can simply use the corresponding Spark class for sort expressions. I did not mind having a custom implementation but it makes equality checks harder.
- Tests for write distribution and ordering.

### CODE REVIEW

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java [hunk 1]
```diff
@@ -148,19 +152,23 @@ public String rewrittenFileSetId() {
   }
 
   public DistributionMode distributionMode() {
-    String defaultValue;
-    if (table.sortOrder().isSorted()) {
-      defaultValue = TableProperties.WRITE_DISTRIBUTION_MODE_RANGE;
-    } else {
-      defaultValue = TableProperties.WRITE_DISTRIBUTION_MODE_DEFAULT;
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:155]: Do we need to continue to respect this? I think the idea was for this to control what to do for unsorted tables, but now this hard-codes NONE. Should we deprecate the property and remove it in 0.14.0?

**Review by rdblue — APPROVED** [review]
Overall looks good.

The only problem is that it now ignores an already public setting, `TableProperties.WRITE_DISTRIBUTION_MODE_DEFAULT`. We should probably continue to use it or deprecate it.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:155]: The problem is that our default value isn't static and cannot be represented by a literal. It is just a constant and we still respect the table property.

I'd probably just deprecate `WRITE_DISTRIBUTION_MODE_DEFAULT`. Do you agree, @rdblue?

Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:155]: Sounds fine to me.

Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:155]: It turns out this property is used by Flink, which makes more sense as Flink does not support `range`. I'll probably keep things as is then.

Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java:155]: I think we should follow up and deprecate it. We can do that in a PR that updates Flink to set its own default like here.

**Pull request merged**